datalog/src/main/java/org/wpilib/datalog/DataLogReaderThread.java

```diff
@@ -0,0 +1,244 @@
+          }
+        } finally {
+          m_mutex.unlock();
+        }
+      } else if (record.isControl()) {
+        System.out.println("Unrecognized control record");
+      }
+    }
+
+    m_done.set(true);
+  }
+
+  /**
+   * Returns the number of records contained in the given DataLog.
+   *
+   * @return The number of records in this DataLog.
+   */
+  public int getNumRecords() {
+    return m_numRecords.get();
+  }
+
+  /**
+   * Returns the number of distinct entries in the DataLog.
+   *
+   * @return The number of distinct entries in the DataLog.
+   */
+  public int getNumEntries() {
+    m_mutex.lock();
+    try {
+      return m_entriesByName.size();
+    } finally {
+      m_mutex.unlock();
+    }
+  }
+
+  /**
+   * Executes a user-provided function on the each uniquely-named entry from the DataLog.
+   *
+   * @param func A function that accepts a <code>DataLogReaderEntry</code> that will be called on
+   *     every entry with a unique name.
+   */
+  public void forEachEntryName(Consumer<DataLogReaderEntry> func) {
+    m_mutex.lock();
+    try {
+      for (var kv : m_entriesByName.entrySet()) {
+        func.accept(kv.getValue());
+      }
+    } finally {
+      m_mutex.unlock();
+    }
+  }
+
+  public boolean isDone() {
+    return m_done.get();
+  }
+
+  public DataLogReader getReader() {
+    return m_reader;
+  }
+
+  /**
+   * Fetches the entry with the given id.
+   *
+   * @param entry Id number of the desired entry, which is associated with all of its records.
+   * @return The DataLogReaderEntry associated with that entry id.
+   */
+  DataLogReaderEntry getEntry(int entry) {
+    m_mutex.lock();
+    try {
+      if (m_entriesById.containsKey(entry)) {
+        return m_entriesById.get(entry);
+      } else {
+        return null;
+      }
+    } finally {
+      m_mutex.unlock();
+    }
+  }
+
+  /**
+   * Fetches the entry with the given name.
+   *
+   * @param name Name string of an entry.
+   * @return The DataLogReaderEntry associated with that name.
+   */
+  DataLogReaderEntry getEntry(String name) {
+    m_mutex.lock();
+    try {
+      if (m_entriesByName.containsKey(name)) {
+        return m_entriesByName.get(name);
+      } else {
+        return null;
+      }
+    } finally {
+      m_mutex.unlock();
+    }
+  }
+
+  private final HashMap<Integer, DataLogReaderEntry> m_entriesById;
+  private final HashMap<String, DataLogReaderEntry> m_entriesByName;
+
+  /**
+   * DataLogReader Entry class, which associates an entry's ID with its name, type, and metadata in
+   * a persistent way.
+   */
+  public static class DataLogReaderEntry extends StartRecordData {
+    private final List<DataLogReaderRange> m_ranges;
+
+    /**
+     * Returns the list of ranges for which this entry is valid.
+     *
+     * @return List of DataLogReaderRange for which this entry is valid
+     */
+    public List<DataLogReaderRange> getRanges() {
+      return m_ranges;
```

Comment:
```suggestion
      return List.copyOf(m_ranges);
```
I'm guessing we don't want clients to modify `m_ranges`.
Another alternative is [`Collections.unmodifiableList()`](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/Collections.html#unmodifiableList(java.util.List)) to have a read-only view that will update as `m_ranges` is updated.

Reply:
Will other code still be able to add to the list and modify elements?